File: src-cpp/leetcode/house-robber.cpp
#include <bits/stdc++.h>
using namespace std;

// iterative dp
int rob(vector<int> &nums)
{
    if (nums.size() == 0)
        return 0;

    int a = 0, b = 0;

    for (auto num : nums)
    {
        int tmp = a;
        a = max(b + num, a);
        b = tmp;
    }

    return a;
}

// recursive memo/dp
int rob(vector<int> &nums)
{
    vector<int> memo(nums.size(), -1);
    function<int(int)> robUtil = [&nums, &memo, &robUtil](int i) -> int
    {
        if (i < 0)
            return 0;

        if (memo[i] >= 0)
            return memo[i];

        memo[i] = max(robUtil(i - 2) + nums[i], robUtil(i - 1));
        return memo[i];
    };

    return robUtil(nums.size() - 1);
}

// naive-2 tle bf
int rob(vector<int> &nums)
{
    function<int(int)> robUtil = [&nums, &robUtil](int i) -> int
    {
        if (i < 0)
            return 0;

        return max(robUtil(i - 2) + nums[i], robUtil(i - 1));
    };

    return robUtil(nums.size() - 1);
}

// naive tle bf
int rob(vector<int> &nums)
{
    function<int(int, int, bool)> robUtil = [&nums, &robUtil](int i, int money, bool canRob) -> int
    {
        if (i == nums.size())
            return money;

        if (canRob)
        {
            return max(robUtil(i + 1, money + nums[i], false), robUtil(i + 1, money, true));
        }

        return robUtil(i + 1, money, true);
    };

    return robUtil(0, 0, true);
}

File: src-ts/leetcode/house-robber.ts
// iterative
function rob(nums: number[]): number {
  let a = 0;
  let b = 0;

  for (const num of nums) {
    [a, b] = [Math.max(b + num, a), a];
  }

  return a;
}

// recursive memo
function robMemo(nums: number[]): number {
  const memoize = (fn: (a: number) => number) => {
    const memo: { [key: number]: number } = {};

    return (i: number) => {
      if (memo[i] !== undefined) return memo[i];

      memo[i] = fn(i);
      return memo[i];
    };
  };

  const robUtilMemo = memoize((i: number): number => {
    if (i < 0) return 0;

    return Math.max(robUtilMemo(i - 2) + nums[i], robUtilMemo(i - 1));
  });

  return robUtilMemo(nums.length - 1);
}

// naive
function robNaive(nums: number[]): number {
  function robUtil(i: number): number {
    if (i < 0) return 0;

    return Math.max(robUtil(i - 2) + nums[i], robUtil(i - 1));
  }

  return robUtil(nums.length - 1);
}
